[BCF-2261] Break up config/v2

cedric-cordenier opened the pull request:
This PR is a proposal to break up the config/v2 package. Before this PR, the package combined multiple responsibilities, namely:
* providing an API to validate structs, via Validate(interface{}).
* providing utilities for dealing with TOML structs, eg. a unified DecodeTOML method.
* the concrete struct types for non-chain-specific config.
* providing an API to deal with env variables.
* providing a toolset to deal with automated doc generation for config settings.

This PR breaks up these responsibilities so that the config/v2 package (renamed config/toml) is only responsible for defining the structs for the non-chain-specific config.
* Move config/env.go into its own package: config/env.
* Move config utilities relating to TOML, ValidateConfig, into utils/configutils.
* Rename config/v2 -> config/toml. This package now only contains the TOML structs for the base config.
* Move config/v2/docs -> config/docs.
* Stop an import cycle by placing the integration tests which verify defaults are correctly extracted for chain-specific configs into a separate docs_test package.

jmank88 commented on core/utils/validate.go:
```diff
@@ -23,7 +21,7 @@ type Validated interface {
 
 // Validate returns any errors from calling Validated.ValidateConfig on cfg and any nested types that implement Validated.
 func Validate(cfg interface{}) (err error) {
```
How about a more specific name if this is going to live in a general utils package?
```suggestion
func ValidateConfig(cfg interface{}) (err error) {
```

cedric-cordenier replied:
If anything, I would suggest making it less, rather than more, specific. That way this can be used generally to validate any struct.

eg.
```
func Validate(s interface{}) error {
...
}

type Validated interface {
     IsValid() error
}
```

jmank88 replied:
But isn't the code hyper-specific to toml config? I'm concerned about moving any pieces over to utils:

> Move config utilities relating to TOML, ValidateConfig, into utils.

It would not be appropriate to use most (all?) of these as-is from other parts of the system - TOML especially, since we use a different module here than for job specs, for example. I think they need to remain config scoped, at least by name, but package separation seems cleaner.

cedric-cordenier replied:
Hmmm... yes, fair point. How about a configutils package containing DecodeTOML, ValidateConfig and the various errors that might be emitted from ValidateConfig() functions?

What I want to achieve with this is to break the dependency between config/v2 and evmconfig/v2, since previously evmconfig/v2 would depend on config/v2, and this caused some import cycles, by moving the common parts out to a separate package.

jmank88 replied:
Actually, in the future LOOP Plugins will be responsible for validating their own config, so we will want to move these generic pieces over to chainlink-relay to be available for other repos. With that in mind, I think it would be simplest to keep a specialized package in the meantime.

jmank88 replied:
> Hmmm... yes, fair point. How about a configutils package containing DecodeTOML, ValidateConfig and the various errors that might be emitted from ValidateConfig() functions?

Yeah I think that would work :+1:

jmank88 replied:
There is a minimal `package config` in chainlink-relay with copies of some error types:
https://github.com/smartcontractkit/chainlink-relay/tree/main/pkg/config

cedric-cordenier replied:
Understood! Would you like me to move these pieces to chainlink-relay and import them here? Or should I just stick to putting them in configutils for now?

jmank88 replied:
I think a local package is fine for now. We can move it later when we move the chain family types for LOOPs.

jmank88 commented on core/chains/cosmos/config.go:
```diff
@@ -17,37 +17,37 @@ import (
 	"github.com/smartcontractkit/chainlink/v2/core/chains"
 	"github.com/smartcontractkit/chainlink/v2/core/chains/cosmos/types"
-	v2 "github.com/smartcontractkit/chainlink/v2/core/config/v2"
+	"github.com/smartcontractkit/chainlink/v2/core/utils/configutils"
```
nit/ I think it would be more idiomatic to name the package just `config` to avoid repeating `utils`:
```suggestion
	"github.com/smartcontractkit/chainlink/v2/core/utils/config"
```
Can still name the imports when there are collisions :shrug: 
```suggestion
	configutils "github.com/smartcontractkit/chainlink/v2/core/utils/config"
```

jmank88 commented on core/config/docs/docs.go:
```diff
@@ -38,12 +38,12 @@ var (
 	exampleSecrets string
 
-	docsTOML = coreTOML + chainsEVMTOML + chainsCosmosTOML + chainsSolanaTOML + chainsStarknetTOML
+	DocsTOML = CoreDefaultsTOML + chainsEVMTOML + chainsCosmosTOML + chainsSolanaTOML + chainsStarknetTOML
```
Is there any way to avoid exporting these vars? Ideally they would be co-located with their usage. IIRC there was an obstacle to using an internal/ directory when I first implemented this, but that may have changed.

jmank88 commented on core/config/env/env.go:
```diff
@@ -0,0 +1,52 @@
+	ThresholdKeyShare            = EnvSecret("CL_THRESHOLD_KEY_SHARE")
+)
+
+type Env string
```
Perhaps this should be called `env.Var` now?
```suggestion
type Var string
```

cedric-cordenier commented on core/config/docs/defaults.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+func CoreDefaults() (c toml.Core) {
+	if (defaults == toml.Core{}) {
```
@jmank88 Let me know what you think of this change; instinctively I think it's an anti-pattern to have packages implement init, so I made the instantiation of `defaults` lazy. This shouldn't change any behaviour since CoreDefaults is still one of the very first calls made at startup.

jmank88 replied:
How is this an improvement over an init func? A `sync.Once` would be more efficient than a mutex, but with the init you don't need to synchronize at all.

jmank88 commented on core/config/docs/docs.go:
```diff
@@ -38,7 +38,7 @@ var (
 	exampleSecrets string
 
-	docsTOML = coreTOML + chainsEVMTOML + chainsCosmosTOML + chainsSolanaTOML + chainsStarknetTOML
+	docsTOML = coreDefaultsTOML + chainsEVMTOML + chainsCosmosTOML + chainsSolanaTOML + chainsStarknetTOML
```
IMHO this is a misleading name change, because they are all defaults.

cedric-cordenier commented:
@jmank88 I've removed my changes to CoreDefault and added the init back in. Your comment about that seems to have disappeared, but my thinking here wasn't primarily about efficiency, it was primarily about the principle of least surprise and avoiding side effects when a package is imported.

jmank88 commented on core/config/env/env.go:
```diff
@@ -0,0 +1,52 @@
+	PrometheusDiscoveryHostName = Var("CL_PROMETHEUS_DISCOVERY_HOSTNAME")
+	// EnvLooopHostName is the hostname used for HTTP communication between the
+	// node and LOOPps. In most cases this does not need to be set explicitly.
+	LooppHostName = Var("CL_LOOPP_HOSTNAME")
```
pre-existing, but since you are touching it:
```suggestion
	LOOPPHostName = Var("CL_LOOPP_HOSTNAME")
```
